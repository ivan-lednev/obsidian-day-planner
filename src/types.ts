import type { Moment } from "moment";
import { MetadataCache, Pos } from "obsidian";
import { DataArray, STask } from "obsidian-dataview";
import { Readable, Writable } from "svelte/store";

import type { getHorizontalPlacing } from "./overlap/horizontal-placing";
import type { ObsidianFacade } from "./service/obsidian-facade";

export interface PlanItemLocation {
  path: string;
  line: number;
  position: Pos;
}

export type OnUpdateFn = (
  baseline: PlanItem[],
  updated: PlanItem[],
) => Promise<void | void[]>;

export interface PlanItem {
  startTime: Moment;
  /**
   * @deprecated this will be replaced with dataview `symbol` and `status`
   */
  listTokens: string;

  // todo: the distinction needs to be clearer
  firstLineText: string;
  text: string;

  durationMinutes: number;
  startMinutes: number;
  location?: PlanItemLocation;
  id: string;
  /**
   * @deprecated derive it from startTime instead
   */
  rawStartTime: string;
  /**
   * @deprecated derive it from startTime and duration instead
   */
  rawEndTime: string;
}

export interface PlacedPlanItem extends PlanItem {
  placing: ReturnType<typeof getHorizontalPlacing>;
  isGhost?: boolean;
}

export type RelationToNow = "past" | "present" | "future";

export type TimeBlock = Pick<
  PlanItem,
  "startMinutes" | "durationMinutes" | "id"
>;

export interface Overlap {
  columns: number;
  span: number;
  start: number;
}

export type Timestamp = {
  startMinutes: number;
  durationMinutes: number;
};

export interface ObsidianContext {
  obsidianFacade: ObsidianFacade;
  metadataCache: MetadataCache;
  onUpdate: OnUpdateFn;
  initWeeklyView: () => Promise<void>;
  dataviewTasks: Readable<DataArray<STask>>;
  refreshTasks: (source: string) => void;
  dataviewLoaded: Writable<boolean>;
  fileSyncInProgress: Readable<boolean>;
}
