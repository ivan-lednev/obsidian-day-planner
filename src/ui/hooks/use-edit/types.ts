import type { Moment } from "moment";

import type { DayPlannerSettings } from "../../../settings";
import type { LocalTask, WithTime } from "../../../task-types";

export enum EditMode {
  DRAG = "DRAG",
  DRAG_AND_SHIFT_OTHERS = "DRAG_AND_SHIFT_OTHERS",
  DRAG_AND_SHRINK_OTHERS = "DRAG_AND_SHRINK_OTHERS",
  RESIZE = "RESIZE",
  RESIZE_FROM_TOP = "RESIZE_FROM_TOP",
  RESIZE_AND_SHIFT_OTHERS = "RESIZE_AND_SHIFT_OTHERS",
  RESIZE_FROM_TOP_AND_SHIFT_OTHERS = "RESIZE_FROM_TOP_AND_SHIFT_OTHERS",
  RESIZE_AND_SHRINK_OTHERS = "RESIZE_AND_SHRINK_OTHERS",
  RESIZE_FROM_TOP_AND_SHRINK_OTHERS = "RESIZE_FROM_TOP_AND_SHRINK_OTHERS",
  CREATE = "CREATE",
  DELETE = "DELETE",
  SCHEDULE_SEARCH_RESULT = "SCHEDULE_SEARCH_RESULT",
}

export interface EditOperation {
  task: WithTime<LocalTask>;
  mode: EditMode;
}

export type TaskTransformer = (
  baseline: WithTime<LocalTask>[],
  editTarget: WithTime<LocalTask>,
  cursorTime: number,
  settings: DayPlannerSettings,
  day?: Moment,
) => WithTime<LocalTask>[];
