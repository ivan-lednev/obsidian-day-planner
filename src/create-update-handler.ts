import type { App } from "obsidian";
import { isNotVoid } from "typed-assert";

import { sortListsRecursivelyInMarkdown } from "./mdast/mdast";
import {
  createTransaction,
  getTaskDiffFromEditState,
  mapTaskDiffToUpdates,
  TransactionWriter,
} from "./service/diff-writer";
import type { PeriodicNotes } from "./service/periodic-notes";
import type { VaultFacade } from "./service/vault-facade";
import type { DayPlannerSettings } from "./settings";
import type { OnUpdateFn } from "./types";
import { type ConfirmationModalProps } from "./ui/confirmation-modal";
import { EditMode } from "./ui/hooks/use-edit/types";
import { SingleSuggestModal } from "./ui/SingleSuggestModal";
import { applyScopedUpdates } from "./util/markdown";

export async function getTextFromUser(app: App): Promise<string | undefined> {
  return new Promise((resolve) => {
    new SingleSuggestModal({
      app,
      getDescriptionText: (value) =>
        value.trim().length === 0
          ? "Start typing to create a task"
          : `Create item "${value}"`,
      onChooseSuggestion: async ({ text }) => {
        resolve(text);
      },
      onClose: () => {
        resolve(undefined);
      },
    }).open();
  });
}

export const createUpdateHandler = (props: {
  settings: () => DayPlannerSettings;
  transactionWriter: TransactionWriter;
  vaultFacade: VaultFacade;
  periodicNotes: PeriodicNotes;
  onEditCanceled: () => void;
  onEditConfirmed: () => void;
  getTextInput: () => Promise<string | undefined>;
  getConfirmationInput: (input: ConfirmationModalProps) => Promise<boolean>;
}): OnUpdateFn => {
  const {
    settings,
    transactionWriter,
    vaultFacade,
    onEditCanceled,
    onEditConfirmed,
    periodicNotes,
    getTextInput,
    getConfirmationInput,
  } = props;

  function getPathsToCreate(paths: string[]) {
    return paths.reduce<string[]>(
      (result, path) =>
        vaultFacade.checkFileExists(path) ? result : result.concat(path),
      [],
    );
  }

  return async (base, next, mode) => {
    const diff = getTaskDiffFromEditState(base, next);

    if (mode === EditMode.CREATE) {
      const created = diff.added[0];

      isNotVoid(created);

      const modalOutput = await getTextInput();

      if (!modalOutput) {
        return onEditCanceled();
      }

      diff.added[0] = { ...created, text: modalOutput };
    }

    const updates = mapTaskDiffToUpdates(diff, mode, settings(), periodicNotes);

    const afterEach = settings().sortTasksInPlanAfterEdit
      ? (contents: string) =>
          applyScopedUpdates(
            contents,
            settings().plannerHeading,
            sortListsRecursivelyInMarkdown,
          )
      : undefined;

    const transaction = createTransaction({
      updates,
      afterEach,
      settings: settings(),
    });

    const updatePaths = [...new Set([...transaction.map(({ path }) => path)])];

    const needToCreate = getPathsToCreate(updatePaths);

    if (needToCreate.length > 0) {
      const confirmed = await getConfirmationInput({
        title: "Need to create files",
        text: `The following files need to be created: ${needToCreate.join("; ")}`,
        cta: "Create",
      });

      if (!confirmed) {
        return onEditCanceled();
      }

      await Promise.all(
        needToCreate.map(async (path) => {
          const date = periodicNotes.getDateFromPath(path, "day");

          isNotVoid(date);

          await periodicNotes.createDailyNote(date);
        }),
      );
    }

    await transactionWriter.writeTransaction(transaction);

    return onEditConfirmed();
  };
};
