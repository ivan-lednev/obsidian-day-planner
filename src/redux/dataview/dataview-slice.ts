import { type PayloadAction } from "@reduxjs/toolkit";
import type { Pos } from "obsidian";
import type { STask } from "obsidian-dataview";

import { createAppSlice } from "../create-app-slice";

export type ListPropsParseResult = {
  parsed: Record<string, unknown>;
  position: Pos;
  raw: string;
};

export type LineToListProps = Record<number, ListPropsParseResult>;

export type PathToListProps = Record<string, LineToListProps>;

interface DataviewSliceState {
  dataviewTasks: Array<STask>;
  dataviewLoaded: boolean;
  listProps: PathToListProps;
}

const initialState: DataviewSliceState = {
  dataviewTasks: [],
  dataviewLoaded: false,
  listProps: {},
};

export const dataviewSlice = createAppSlice({
  name: "dataview",
  initialState,
  reducers: (create) => ({
    dataviewChange: create.reducer((state, action: PayloadAction<string>) => {
      state.dataviewLoaded = true;
    }),
    dataviewTasksUpdated: create.reducer(
      (state, action: PayloadAction<Array<STask>>) => {
        state.dataviewTasks = action.payload;
      },
    ),
    listPropsParsed: create.reducer(
      (
        state,
        action: PayloadAction<{
          path: string;
          lineToListProps?: LineToListProps;
        }>,
      ) => {
        const { path, lineToListProps } = action.payload;

        if (!lineToListProps || Object.keys(lineToListProps).length === 0) {
          return;
        }

        state.listProps[path] = lineToListProps;
      },
    ),
  }),
  selectors: {
    selectDataviewTasks: (state) => state.dataviewTasks,
    selectListProps: (state) => state.listProps,
    selectListPropsForPath: (state, path: string) => state.listProps[path],
    selectDataviewLoaded: (state) => state.dataviewLoaded,
  },
});

export const { dataviewChange, dataviewTasksUpdated, listPropsParsed } =
  dataviewSlice.actions;

export const {
  selectDataviewTasks,
  selectListProps,
  selectDataviewLoaded,
  selectListPropsForPath,
} = dataviewSlice.selectors;

export type DataviewChangeAction = ReturnType<typeof dataviewChange>;
