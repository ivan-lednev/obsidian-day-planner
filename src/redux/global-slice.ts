import { createSelector, type PayloadAction } from "@reduxjs/toolkit";

import { defaultDayFormat } from "../constants";

import { createAppSlice } from "./create-app-slice";

interface ObsidianSliceState {
  visibleDays: string[];
}

export const initialState: ObsidianSliceState = {
  visibleDays: [],
};

export const globalSlice = createAppSlice({
  name: "obsidian",
  initialState,
  reducers: (create) => ({
    visibleDaysUpdated: create.reducer(
      (state, action: PayloadAction<string[]>) => {
        state.visibleDays = action.payload;
      },
    ),
    editCanceled: () => {},
  }),
  selectors: {
    selectVisibleDays: (state) => state.visibleDays,
  },
});

export const { visibleDaysUpdated, editCanceled } = globalSlice.actions;

export const { selectVisibleDays } = globalSlice.selectors;

export const selectSortedDedupedVisibleDays = createSelector(
  selectVisibleDays,
  (days) => {
    return [...new Set(days)]
      .sort()
      .map((it) => window.moment(it, defaultDayFormat));
  },
);
