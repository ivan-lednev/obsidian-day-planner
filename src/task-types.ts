import type { Moment } from "moment";
import type { AttendeePartStat } from "node-ical";
import type { Pos } from "obsidian";

import type { HorizontalPlacing } from "./overlap/horizontal-placing";
import type { IcalConfig } from "./settings";
import type { Props } from "./util/props";

export interface Entry {
  type: "listItem" | "task" | "icalEvent" | "file";
  props: Props;
}

export interface TaskLocation {
  path: string;
  position: Pos;
}

export interface TaskTokens {
  symbol: string;
  status?: string;
}

export interface FileLine {
  text: string;
  line: number;
  task: boolean;
}

export type WithPlacing<T> = T & {
  placing: HorizontalPlacing;
};

export type BaseTask = {
  /**
   * Tasks get an ID on parsing. It is unique to a line in a file, not to a
   * block, visible in the UI (because blocks might get split at midnight, etc.).
   */
  id: string;
  startTime: Moment;
  isAllDayEvent?: boolean;
  // TODO: move to TimeBlockView
  truncated?: Side[];
};

export type WithTime<T> = T & {
  durationMinutes: number;
};

export interface RemoteTask extends BaseTask {
  calendar: IcalConfig;
  summary: string;
  rsvpStatus: AttendeePartStat;
  description?: string;
  location?: string;
}

type Side = "top" | "bottom" | "left" | "right";

export interface LocalTask extends TaskTokens, BaseTask {
  text: string;
  lines?: Array<FileLine>;

  // todo: move out to InMemoryTask
  location?: TaskLocation;

  // todo: move to Time
  durationMinutes: number;
}

export type TaskWithoutComputedDuration = Omit<LocalTask, "durationMinutes"> &
  Partial<Pick<LocalTask, "durationMinutes">>;

export type Task = LocalTask | RemoteTask;

export function isRemote<T extends Task>(task: T): task is T & RemoteTask {
  return Object.hasOwn(task, "calendar");
}

export function isLocal(task: Task): task is LocalTask {
  return Object.hasOwn(task, "text");
}
