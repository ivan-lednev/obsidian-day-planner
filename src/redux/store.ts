import type {
  Action,
  ConfigureStoreOptions,
  ListenerEffect,
  ListenerMiddlewareInstance,
  ThunkAction,
  TypedStartListening,
} from "@reduxjs/toolkit";
import { combineSlices, configureStore } from "@reduxjs/toolkit";
import type { MetadataCache, TFile, Vault } from "obsidian";

import { icalRefreshIntervalMillis } from "../constants";
import type DayPlanner from "../main";
import type { DataviewFacade } from "../service/dataview-facade";
import type { PluginData } from "../settings";
import type { ReduxExtraArgument } from "../types";

import { dataviewChange, dataviewSlice } from "./dataview/dataview-slice";
import { globalSlice } from "./global-slice";
import {
  icalRefreshRequested,
  icalSlice,
  type IcalState,
  initialIcalState,
} from "./ical/ical-slice";
import { initListenerMiddleware } from "./listener-middleware";
import { searchSlice } from "./search-slice";
import { settingsSlice } from "./settings-slice";

const rootReducer = combineSlices(
  globalSlice,
  searchSlice,
  dataviewSlice,
  settingsSlice,
  icalSlice,
);

export type RootState = ReturnType<typeof rootReducer>;

export const makeStore = (
  props: Omit<
    ConfigureStoreOptions<RootState>,
    "preloadedState" | "reducer"
  > & {
    preloadedState?: Partial<RootState>;
  },
) => {
  const { preloadedState, middleware } = props;

  return configureStore({
    reducer: rootReducer,
    middleware,
    preloadedState,
  });
};

export function initStoreForPlugin(props: {
  pluginData: PluginData;
  dataviewFacade: DataviewFacade;
  vault: Vault;
  metadataCache: MetadataCache;
  plugin: DayPlanner;
}) {
  const { pluginData, dataviewFacade, vault, metadataCache, plugin } = props;

  const icalStateWithCachedRawIcals: IcalState = {
    ...initialIcalState,
    plainTextIcals: pluginData.rawIcals || [],
  };

  // todo: extra is not needed at all
  const listenerMiddleware = initListenerMiddleware({
    extra: {
      dataviewFacade,
      vault,
      metadataCache,
      onIcalsFetched: async (rawIcals) => {
        await plugin.saveData({ ...plugin.settings(), rawIcals });
      },
    },
  });

  const store = makeStore({
    preloadedState: {
      ical: icalStateWithCachedRawIcals,
    },
    middleware: (getDefaultMiddleware) => {
      return getDefaultMiddleware().concat(listenerMiddleware.middleware);
    },
  });

  plugin.register(() => {
    listenerMiddleware.clearListeners();
  });

  plugin.registerInterval(
    window.setInterval(() => {
      store.dispatch(icalRefreshRequested());
    }, icalRefreshIntervalMillis),
  );

  plugin.registerEvent(
    metadataCache.on(
      // @ts-expect-error
      "dataview:metadata-change",
      (eventType: unknown, file: TFile) =>
        store.dispatch(dataviewChange(file.path)),
    ),
  );

  return { store, listenerMiddleware };
}

export type AppStore = ReturnType<typeof makeStore>;
export type AppDispatch = AppStore["dispatch"];
export type AppThunk<ThunkReturnType = void> = ThunkAction<
  ThunkReturnType,
  RootState,
  unknown,
  Action
>;
export type StartListeningFn = TypedStartListening<
  RootState,
  AppDispatch,
  ReduxExtraArgument
>;
export type AppListenerEffect<A extends Action = Action> = ListenerEffect<
  A,
  RootState,
  AppDispatch
>;
export type AppListenerMiddlewareInstance = ListenerMiddlewareInstance<
  RootState,
  AppDispatch,
  ReduxExtraArgument
>;
