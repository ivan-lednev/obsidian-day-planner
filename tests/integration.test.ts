import { readdir, readFile, stat } from "node:fs/promises";
import path from "node:path";

import { noop } from "lodash/fp";
import type { Moment } from "moment";
import { type CachedMetadata, MetadataCache, type Vault } from "obsidian";
import type { SListEntry, STask } from "obsidian-dataview";
import { derived, get, writable } from "svelte/store";
import { isNotVoid } from "typed-assert";
import { describe, expect, test, vi } from "vitest";

import { createUpdateHandler } from "../src/create-update-handler";
import { dataviewChange } from "../src/redux/dataview/dataview-slice";
import { initialState } from "../src/redux/global-slice";
import { createReactor, type RootState } from "../src/redux/store";
import type { DataviewFacade } from "../src/service/dataview-facade";
import { TransactionWriter } from "../src/service/diff-writer";
import type { PeriodicNotes } from "../src/service/periodic-notes";
import { VaultFacade } from "../src/service/vault-facade";
import type { WorkspaceFacade } from "../src/service/workspace-facade";
import {
  type DayPlannerSettings,
  defaultSettingsForTests,
} from "../src/settings";
import { isLocal, type Task } from "../src/task-types";
import { EditMode } from "../src/ui/hooks/use-edit/types";
import { useTasks } from "../src/ui/hooks/use-tasks";
import { getOneLineSummary } from "../src/util/task-utils";

import {
  createInMemoryFile,
  FakeDataviewFacade,
  FakeMetadataCache,
  FakePeriodicNotes,
  FakeWorkspaceFacade,
  getPathToDiff,
  InMemoryVault,
  type InMemoryFile,
} from "./test-utils";

const { join } = path.posix;

const dailyNoteFileNames = ["2025-07-19", "2025-07-20", "2025-07-28"];

const fixturesDirPath = "fixtures";
const dumpPath = join(fixturesDirPath, "metadata-dump", "tasks.json");
const fixtureVaultPath = join(fixturesDirPath, "fixture-vault");

async function loadMetadataDump() {
  const files = await readdir(fixtureVaultPath);

  const inMemoryFiles = await Promise.all(
    files.map(async (file) => {
      const filePath = join(fixtureVaultPath, file);

      const stats = await stat(filePath);

      if (!stats.isFile()) {
        throw new TypeError(
          `Only files are supported in fixture vault, not this: ${filePath}`,
        );
      }

      const contents = await readFile(filePath, "utf8");

      return createInMemoryFile({ path: filePath, contents });
    }),
  );

  const rawMetadataDump = await readFile(dumpPath, "utf-8");

  const metadataDump = JSON.parse(rawMetadataDump);
  const { tasks, lists, cachedMetadata } = metadataDump;

  const pathToInMemoryFile = Object.fromEntries(
    inMemoryFiles.map((it) => [it.path, it]),
  );

  const inMemoryDailyNotes = dailyNoteFileNames.map((it) => {
    const path = join(fixtureVaultPath, `${it}.md`);
    const file = pathToInMemoryFile[path];

    isNotVoid(file, `There is no file for key: '${it}'`);

    return { path, file, date: window.moment(it) };
  });

  return {
    inMemoryFiles,
    inMemoryDailyNotes,
    tasks,
    lists,
    cachedMetadata,
  };
}

function initTestServices(props: {
  inMemoryFiles: InMemoryFile[];
  inMemoryDailyNotes: { path: string; file: InMemoryFile; date: Moment }[];
  tasks: STask[];
  lists: SListEntry[];
  cachedMetadata: Record<string, CachedMetadata>;
}) {
  const { inMemoryFiles, inMemoryDailyNotes, tasks, lists, cachedMetadata } =
    props;

  const periodicNotes = new FakePeriodicNotes(
    inMemoryDailyNotes,
  ) as unknown as PeriodicNotes;

  const dataviewFacade = new FakeDataviewFacade({
    tasks,
    lists,
  }) as unknown as DataviewFacade;

  const metadataCache = new FakeMetadataCache(
    cachedMetadata,
  ) as unknown as MetadataCache;

  const vault = new InMemoryVault(inMemoryFiles);

  const getTasksApi = () => {
    throw new Error("Can't access tasks API inside tests");
  };

  const vaultFacade = new VaultFacade(vault as unknown as Vault, getTasksApi);

  const transactionWriter = new TransactionWriter(vaultFacade);

  const workspaceFacade =
    new FakeWorkspaceFacade() as unknown as WorkspaceFacade;

  return {
    periodicNotes,
    dataviewFacade,
    metadataCache,
    vault,
    transactionWriter,
    workspaceFacade,
    vaultFacade,
  };
}

async function setUp(props: {
  visibleDays: string[];
  settings?: DayPlannerSettings;
}) {
  const { visibleDays, settings = defaultSettingsForTests } = props;

  const { inMemoryFiles, inMemoryDailyNotes, tasks, lists, cachedMetadata } =
    await loadMetadataDump();

  const {
    periodicNotes,
    dataviewFacade,
    metadataCache,
    vault,
    transactionWriter,
    workspaceFacade,
    vaultFacade,
  } = initTestServices({
    inMemoryFiles,
    inMemoryDailyNotes,
    tasks,
    lists,
    cachedMetadata,
  });

  const visibleDaysStore = writable(visibleDays.map((it) => window.moment(it)));
  const isOnline = writable(true);
  const layoutReady = writable(true);
  const settingsStore = writable(settings);
  const currentTime = writable(window.moment());

  const onEditCanceled = vi.fn();
  const onEditConfirmed = vi.fn();
  const onIcalsFetched = vi.fn();

  const defaultPreloadedStateForTests: Partial<RootState> = {
    obsidian: {
      ...initialState,
      visibleDays,
    },
    settings: { settings },
  };

  const {
    dispatch,
    remoteTasks,
    taskUpdateTrigger,
    listProps,
    pointerDateTime,
  } = createReactor({
    preloadedState: {
      ...defaultPreloadedStateForTests,
      obsidian: {
        ...initialState,
        visibleDays,
      },
    },
    dataviewFacade,
    metadataCache,
    vault: vault as unknown as Vault,
    onIcalsFetched,
  });

  inMemoryFiles.forEach(({ path }) => {
    dispatch(dataviewChange(path));
  });

  const onUpdate = createUpdateHandler({
    settings: () => settings,
    transactionWriter,
    vaultFacade,
    periodicNotes,
    onEditCanceled,
    onEditConfirmed,
    getTextInput: () => Promise.resolve("Text input"),
    getConfirmationInput: () => Promise.resolve(true),
  });

  const {
    tasksWithActiveClockProps,
    getDisplayedTasksWithClocksForTimeline,
    tasksWithTimeForToday,
    editContext,
    newlyStartedTasks,
  } = useTasks({
    onUpdate,
    onEditAborted: () => {},
    periodicNotes,
    dataviewFacade,
    metadataCache,
    workspaceFacade,
    isOnline,
    visibleDays: visibleDaysStore,
    layoutReady,
    debouncedTaskUpdateTrigger: taskUpdateTrigger,
    dataviewChange: taskUpdateTrigger,
    settingsStore,
    currentTime,
    pointerDateTime,
    remoteTasks,
    listProps,
  });

  const allTasks = derived(
    editContext.dayToDisplayedTasks,
    ($dayToDisplayedTasks) => {
      return Object.values($dayToDisplayedTasks).flatMap(
        ({ withTime, noTime }) => withTime.concat(noTime),
      );
    },
  );

  // this prevents the store from resetting;
  allTasks.subscribe(noop);

  function moveCursorTo(
    dateTime: Moment,
    type: "date" | "dateTime" = "dateTime",
  ) {
    pointerDateTime.set({
      dateTime,
      type,
    });
  }

  function findTask(predicate: (task: Task) => boolean) {
    const found = get(allTasks).filter(isLocal).find(predicate);

    isNotVoid(found, `Task not found`);

    return found;
  }

  function findByText(text: string) {
    return findTask((it) => getOneLineSummary(it).includes(text));
  }

  await vi.waitFor(() => {
    const areTasksLoaded = get(allTasks).length > 0;
    const areClocksLoaded = get(tasksWithActiveClockProps).length > 0;

    expect(areTasksLoaded || areClocksLoaded).toBeTruthy();
  });

  return {
    dispatch,
    tasksWithActiveClockProps,
    getDisplayedTasksWithClocksForTimeline,
    tasksWithTimeForToday,
    editContext,
    newlyStartedTasks,
    periodicNotes,
    moveCursorTo,
    onUpdate,
    vault,
    findTask,
    findByText,
    allTasks,
    transactionWriter,
    currentTime,
  };
}

describe("Clocks", () => {
  test("Reads log data", async () => {
    const { getDisplayedTasksWithClocksForTimeline } = await setUp({
      visibleDays: ["2025-07-19"],
    });

    const displayedTasks = getDisplayedTasksWithClocksForTimeline(
      window.moment("2025-07-19"),
    );

    expect(get(displayedTasks)).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          startTime: window.moment("2025-07-19 12:00"),
          durationMinutes: 150,
        }),
        expect.objectContaining({
          startTime: window.moment("2025-07-19 15:00"),
          durationMinutes: 90,
        }),
        expect.objectContaining({
          startTime: window.moment("2025-07-19 13:00"),
          durationMinutes: 210,
        }),
      ]),
    );
  });

  test.todo("Ignores invalid dates, extra keys");

  test("Tasks with active clocks receive durations set to current time", async () => {
    const {
      getDisplayedTasksWithClocksForTimeline,
      currentTime,
      tasksWithActiveClockProps,
    } = await setUp({
      visibleDays: ["2025-01-11"],
    });

    currentTime.set(window.moment("2025-01-01 17:30"));

    const displayedClocks = getDisplayedTasksWithClocksForTimeline(
      window.moment("2025-01-01"),
    );

    expect(get(displayedClocks)).toMatchObject([
      {
        startTime: window.moment("2025-01-01 17:00"),
        durationMinutes: 30,
      },
      {
        startTime: window.moment("2025-01-01 13:00"),
        durationMinutes: 120,
      },
    ]);

    expect(get(tasksWithActiveClockProps)).toMatchObject([
      {
        startTime: window.moment("2025-01-01 17:00"),
        durationMinutes: 30,
      },
    ]);
  });

  test.todo("Splits log entries over midnight");

  test.todo("Parses code blocks under nested tasks");

  describe("Clocking in", () => {
    test.todo("Clocks in on tasks");

    test.todo("Does not clock in on tasks with active clocks");

    test.todo("Does not mess up other child blocks in list");

    test.todo(
      "For a single line in a file, adds a newline instead of appending the code block to the line",
    );
  });

  describe("Clocking out", () => {
    test.todo("Clocks out on tasks");

    test.todo("Does not clock out on tasks without a clock");
  });

  describe("Canceling clocks", () => {
    test.todo("Cancels clocks");

    test.todo("Does not touch a file without an active clock");
  });
});

describe("Task views", () => {
  describe("Frontmatter", () => {
    test.todo("Shows log entries from frontmatter");

    test.todo("Edits log entries from frontmatter");
  });

  test.fails(
    "Ignores tasks and lists outside of planner section in daily notes",
    async () => {
      const { editContext } = await setUp({
        visibleDays: ["2025-07-19"],
      });

      const displayedTasks = editContext.getDisplayedTasksForTimeline(
        window.moment("2025-07-19"),
      );

      expect(get(displayedTasks)?.withTime).toHaveLength(1);
      expect(get(displayedTasks)?.withTime).toMatchObject([
        { startTime: window.moment("2025-07-19 11:00"), durationMinutes: 30 },
      ]);
    },
  );

  test.todo("Tasks do not contain duplicates");

  test.todo("Combines tasks from daily notes with tasks from other files");
});

describe("Editing", () => {
  describe("Daily notes", () => {
    test("Edits tasks", async () => {
      const { editContext, moveCursorTo, vault, findByText } = await setUp({
        visibleDays: ["2025-07-19"],
      });

      editContext.handlers.handleGripMouseDown(
        findByText("List item under planner heading"),
        EditMode.DRAG,
      );

      moveCursorTo(window.moment("2025-07-19 17:00"));

      await editContext.confirmEdit();

      expect(getPathToDiff(vault.initialState, vault.state)).toMatchSnapshot();
    });

    test("Un-schedules tasks", async () => {
      const { editContext, moveCursorTo, vault, findByText } = await setUp({
        visibleDays: ["2025-07-19"],
      });

      editContext.handlers.handleGripMouseDown(
        findByText("List item under planner heading"),
        EditMode.DRAG,
      );

      moveCursorTo(window.moment("2025-07-19"), "date");

      await editContext.confirmEdit();

      expect(getPathToDiff(vault.initialState, vault.state)).toMatchSnapshot();
    });

    test("Creates tasks", async () => {
      const { editContext, moveCursorTo, vault } = await setUp({
        visibleDays: ["2025-07-19", "2025-07-20"],
      });

      moveCursorTo(window.moment("2025-07-20 13:00"));
      editContext.handlers.handleContainerMouseDown();
      moveCursorTo(window.moment("2025-07-20 14:00"));

      await editContext.confirmEdit();

      expect(getPathToDiff(vault.initialState, vault.state)).toMatchSnapshot();
    });

    test(`* Moves a nested task with text between notes
* Does not touch invalid markdown
* Undoes the move`, async () => {
      const {
        editContext,
        moveCursorTo,
        vault,
        findByText,
        transactionWriter,
      } = await setUp({
        visibleDays: ["2025-07-28"],
        settings: {
          ...defaultSettingsForTests,
          sortTasksInPlanAfterEdit: true,
        },
      });

      editContext.handlers.handleGripMouseDown(
        findByText("Child"),
        EditMode.DRAG,
      );

      moveCursorTo(window.moment("2025-07-20 17:00"));

      await editContext.confirmEdit();

      expect(getPathToDiff(vault.initialState, vault.state)).toMatchSnapshot();

      await transactionWriter.undo();

      expect(
        Object.keys(getPathToDiff(vault.initialState, vault.state)).length,
      ).toBe(0);
    });

    describe("Sorting by time", () => {
      test.todo("Sorts tasks after non-mdast edit");

      test.todo("Skips sorting if day planner heading is not found");
    });
  });

  describe("Obsidian-tasks", () => {
    test("Schedules tasks & un-schedules tasks", async () => {
      const { editContext, moveCursorTo, vault, findByText } = await setUp({
        visibleDays: ["2025-07-19"],
      });

      editContext.handlers.handleGripMouseDown(
        findByText("Task without time"),
        EditMode.DRAG,
      );

      moveCursorTo(window.moment("2025-07-19 13:00"));

      await editContext.confirmEdit();

      editContext.handlers.handleGripMouseDown(
        findByText("Task with time"),
        EditMode.DRAG,
      );

      moveCursorTo(window.moment("2025-07-19"), "date");

      await editContext.confirmEdit();

      expect(getPathToDiff(vault.initialState, vault.state)).toMatchSnapshot();
    });

    test.todo(
      "Updates tasks plugin props without duplicating timestamps if moved to same time on another day",
    );
  });
});
