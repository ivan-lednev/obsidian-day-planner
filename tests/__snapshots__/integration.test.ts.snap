// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Editing > Daily notes > Creates tasks 1`] = `
{
  "fixtures/fixture-vault/2025-07-20.md": "
+ # Day planner
+ 
+ - [ ] 13:00 - 14:00 Text input
",
}
`;

exports[`Editing > Daily notes > Edits tasks 1`] = `
{
  "fixtures/fixture-vault/2025-07-19.md": "
- - 11:00 - 11:30 List item under planner heading
+ - 17:00 - 17:30 List item under planner heading
",
}
`;

exports[`Editing > Daily notes > Un-schedules tasks 1`] = `
{
  "fixtures/fixture-vault/2025-07-19.md": "
- - 11:00 - 11:30 List item under planner heading
+ - List item under planner heading
",
}
`;

exports[`Editing > Obsidian-tasks > Schedules tasks & un-schedules tasks 1`] = `
{
  "fixtures/fixture-vault/tasks.md": "
- - [ ] Task without time ⏳ 2025-07-19
+ - [ ] 13:00 - 13:30 Task without time ⏳ 2025-07-19
...
- - [ ] 10:00 - 10:30 Task with time ⏳ 2025-07-19
+ - [ ] Task with time ⏳ 2025-07-19
",
}
`;
